src/wrapper/java_vm/vm.rs

```diff
@@ -49,45 +52,88 @@ impl JavaVM {
     /// when dropped. Calling this for a thread that is already attached is a no-op.
+    ///
+    /// Attached threads [block JVM exit][block].
+    ///
+    /// Attaching and detaching is time-consuming operation, therefore multiple short-term attach
+    /// operations on the same thread should be avoided or replaced with
+    /// `attach_current_thread_permanently`.
+    ///
+    /// [block]: https://docs.oracle.com/en/java/javase/12/docs/specs/jni/invocation.html#unloading-the-vm
     pub fn attach_current_thread(&self) -> Result<AttachGuard> {
-        let (env, requires_detach) = match self.get_env() {
-            Ok(env) => (env, false),
-            Err(_) => unsafe {
-                let mut ptr = ptr::null_mut();
-                let res = java_vm_unchecked!(self.0, AttachCurrentThread, &mut ptr, ptr::null_mut());
-                jni_error_code_to_result(res)?;
-
-                let env = JNIEnv::from_raw(ptr as *mut sys::JNIEnv)?;
-                (env, true)
+        match self.get_env() {
+            Ok(env) => {
+                Ok(AttachGuard::new_nested(env))
             },
-        };
+            Err(_) => {
+                let env_ptr = InternalAttachGuard::create_and_attach(
+                    self.get_java_vm_pointer(),
+                    ThreadType::Normal
+                )?;
+                let env = unsafe {
+                    JNIEnv::from_raw(env_ptr)?
+                };
+                Ok(AttachGuard::new(env))
+            },
+        }
+    }
 
-        Ok(AttachGuard {
-            java_vm: self,
-            env,
-            requires_detach,
-        })
+    /// Detaches current thread from the JVM.
+    ///
+    /// Detaching a non-attached thread is no-op.
+    ///
+    /// Calling this method is an equivalent for calling `drop()` for `AttachGuard`.
+    pub fn detach_current_thread(&self) {
+        InternalAttachGuard::clear_tls();
     }
 
     /// Attaches the current thread to a Java VM as a daemon.
+    ///
+    /// Thread will be automatically detached when finished.
     pub fn attach_current_thread_as_daemon(&self) -> Result<JNIEnv> {
-        let mut ptr = ptr::null_mut();
-        unsafe {
-            let res = java_vm_unchecked!(
-                self.0,
-                AttachCurrentThreadAsDaemon,
-                &mut ptr,
-                ptr::null_mut()
-            );
-            jni_error_code_to_result(res)?;
-
-            JNIEnv::from_raw(ptr as *mut sys::JNIEnv)
+        match self.get_env() {
+            Ok(env) => Ok(env),
+            Err(_) => {
+                let env_ptr = InternalAttachGuard::create_and_attach(
+                    self.get_java_vm_pointer(),
+                    ThreadType::Daemon
+                )?;
+                unsafe { JNIEnv::from_raw(env_ptr) }
+            }
         }
     }
 
+    /// Returns current number of threads attached to the JVM.
+    pub fn threads_attached(&self) -> usize {
+        ATTACHED_THREADS.load(Ordering::SeqCst)
```

Comment:
Just curious: is it only needed for debug purpose or do you build some logic on top of this?

Reply:
I think for debug/diagnostic purposes + easier testing. I considered to add that to documentation separately.

BTW, we can add a fancier test that ensures that a thread is _indeed_ detached: in a loop, create a thread, attach permanently, allocate a java byte array and leak it (= don't use auto_locals or local_frame), join. Create that many threads so that the total allocated Java heap exceeds the configured limit.